core/src/hir/attrs.rs

```diff
@@ -565,6 +592,12 @@ impl Attrs {
 #[non_exhaustive]
 #[derive(Copy, Clone, Debug, Default)]
 pub struct BackendAttrSupport {
+    /// Renaming types/methods, usually to make them more idiomatic.
+    ///
+    /// This is supported by all backends *except for C*.
+    pub renaming: bool,
+    /// Namespacing types, e.g. C++ `namespace`.
+    pub namespacing: bool,
```

Comment:
I don't see a use case for gating on these. All backends support renaming, and there's no alternative action to namespacing.

Reply:
Namespacing is the most important one? If a backend doesn't support namespacing you may rename things.

`#[diplomat::attr(not(supports = namespacing), rename = "ICU4X{0})]"` is great.

Agree on renaming.

Reply:
The backend should just support namespacing. Slapping an `ICU4X` prefix on everything in all languages is not going to be idiomatic. In C for example it should be `ICU4X_`, which would be the namespace implementation for C.

Reply:
No, this is explicitly counter to a goal of attributes which is to enable backends to support fewer things if they choose to, either during their development or long term. "everyone should support namespacing" is a potentially valid thing to be discussed but it is a new ask of backends.

Also there are two kinds of namespacing, JS/Dart come bundled with a default package namespace so it's much less necessary than, say, C++ where things splat into the main namespace by default. It's not even clear to me if JS/Dart need to get namespaces soon.

(we should introduce a separate attribute for distinguishing C++-style needs-namespacing from js-style namespacing-for-code-organization)

Reply:
WIP backends can just ignore the attribute. What I'm saying is that renaming is not a patch for namespacing, a backend should think about what namespacing means in its language, and implement it correctly (at some point). Clients should not be able to say "for any backend that doesn't have namespacing, or hasn't implemented it yet, I'll just do some renames". They are already able to say "for a specific language that doesn't have namespacing, i.e. C, I'll do some renames that make sense in that language".

Reply:
I don't want code that wants to work with such backends to have to manually list backends whilst gating. Again, this is the whole point of the attribute system. A "WIP" backend isn't just when it's still being cobbled together, I expect a lot of backends to be not-fully-feature-complete-but-usable for a while. We're already in that case for the kotlin backend.

Anyway, perhaps review https://github.com/rust-diplomat/diplomat/pull/585 ? This implements the change I was suggesting for #563, which I approved with the understanding that we'd eventually move to a model like this but I didn't want to hinder your work. I'm happy to do the ICU4X-side refactor that that implies.

I don't understand the downside of allowing this fine-grained gating.

Reply:
If you're using renaming to do namespacing you should manually list the backend, because backends have very different naming conventions. That's all I'm saying. Letting users gate on `supports = namespacing` allows renaming as a poor-man's namespacing, but that's just wrong.

Reply:
No, because you might use `supports = namespacing` for code organization as well, making multiple different namespaces / etc, and I don't think renaming is necessarily a "poor man's namespacing" there, it's just a different way of doing things.

Reply:
Either way this is all assuming every backend gets namespacing, and we currently have not resourced that work, and there are still open questions about that.

I really don't see a strong reason for disallowing this kind of gating. I think we should support gating in as many ways as possible, and if people misuse it, that's up to them, I don't think we need to predecide what will and won't be idiomatic.